File: 125_algorithms/_exercises/templates/_algorithms_challenges/leetcode/leetCode/BinarySearch/69_Sqrt_x.py
#! /usr/bin/env python
# -*- coding: utf-8 -*-


c.. Solution o..
    # Binary search.
    ___ mySqrt  x
        low, high = 0, x
        _____ low <= high:
            mid = (low + high) / 2
            __ mid ** 2 <= x < (mid + 1) ** 2:
                r_ mid
            ____ mid ** 2 > x:
                high = mid - 1
            ____
                low = mid + 1


c.. Solution_2 o..
    # Newton iterative method
    # According to:
    # http://www.matrix67.com/blog/archives/361
    ___ mySqrt  x
        __ n.. x:
            r_ x
        val = x
        sqrt_x = (val + x * 1.0 / val) / 2.0
        _____ val - sqrt_x > 0.001:
            val = sqrt_x
            sqrt_x = (val + x * 1.0 / val) / 2.0

        r_ int(sqrt_x)


c.. Solution_3 o..
    # Shorter Newton method.
    ___ mySqrt  x
        val = x
        _____ val * val > x:
            val = (val + x / val) / 2
        r_ val

"""
0
1
15
90
1010
"""

File: 125_algorithms/_exercises/templates/_algorithms_challenges/leetcode/leetCode/BinarySearch/69_Sqrt_x.cpp
/*
 * @Author: xuezaigds@gmail.com
 * @Last Modified time: 2016-07-01 20:42:32
 */

c.. Solution {
public:
    int mySqrt(int x) {
        __(x < 2){
            r_ x;
        }
        int low = 0, high = x;
        _____(low <= high){
            int mid = low + (high - low) / 2;
            // mid * mid will overflow when mid > sqrt(INT_MAX)
            __(x / mid >= mid  && x / (mid+1) < (mid+1)){
                r_ mid;
            }
            else __(x / mid < mid){
                high = mid - 1;
            }
            else{
                low = mid + 1;
            }
        }
        r_ -1; // Can'e be reached!
    }
};

c.. Solution_2{
public:
    int mySqrt(int x) {
        // Newton iterative method
        long ans = x;
        _____(ans * ans > x){
            ans = (ans + x/ans) / 2;
        }
        r_ ans;
    }
};

/*
0
1
15
90
1010
*/
